sal-service/src/main/java/org/ow2/proactive/sal/service/service/CloudService.java

```diff
@@ -239,17 +243,25 @@ public Boolean removeClouds(String sessionId, List<String> cloudIds, Boolean pre
                              cloud.getCloudType());
             }
             if (Boolean.TRUE.equals(tempFlag)) {
+                List<Deployment> toBeRemovedDeployments = new ArrayList<>();
                 if (cloud.getDeployments() != null) {
-                    LOGGER.info("Cleaning deployments from related tasks " + cloud.getDeployments().toString());
-                    cloud.getDeployments().forEach(deployment -> deployment.getTask().removeDeployment(deployment));
+                    LOGGER.info("Cleaning all deployments related to cloud \"{}\"", cloud.getCloudId());
+                    toBeRemovedDeployments.addAll(cloud.getDeployments());
+                    for (Deployment deployment : toBeRemovedDeployments) {
+                        repositoryService.deleteDeployment(deployment);
+                    }
                 }
-                LOGGER.info("Cleaning deployments from the cloud entry");
-                cloud.clearDeployments();
                 LOGGER.info("Cleaning node candidates");
+                List<String> toBeRemovedCloud = Collections.singletonList(cloud.getCloudId());
                 try {
-                    ASYNC_NODE_CANDIDATES_PROCESSES_RESULTS.add(updatingNodeCandidatesUtils.asyncClean(Collections.singletonList(cloud.getCloudId())));
-                } catch (InterruptedException ie) {
-                    LOGGER.warn("Thread cleaning node candidates interrupted!", ie);
+                    long cleaned = nodeCandidateUtils.cleanNodeCandidates(toBeRemovedCloud);
+                    LOGGER.info("Cleaning node candidates related to clouds {} ended properly with {} NC cleaned.",
+                                toBeRemovedCloud.toString(),
```

Comment:
`LOGGER.info("Cleaning node candidates related to clouds {} ended properly with {} NC cleaned.", cloud.getCloudId())`

Reply:
(and same just below)